src/esp/core/Configuration.cpp

```diff
@@ -0,0 +1,246 @@
+      break;
+    case ConfigStoredType::MagnumVec3:
+      v.~Vector3();
+      break;
+    case ConfigStoredType::MagnumQuat:
+      q.~Quaternion();
+      break;
+    case ConfigStoredType::MagnumRad:
+      r.~Rad();
+      break;
+    default:
+      ESP_CHECK(true, "Attempted to delete unknown/unsupported type :" + src);
+  }  // switch
+  type = ConfigStoredType::Unknown;
+}
+
+bool ConfigValue::set(int _i) {
+  checkTypeAndDest(ConfigStoredType::Integer);
+  i = _i;
+  type = ConfigStoredType::Integer;
+  return true;
+}
+
+bool ConfigValue::set(bool _b) {
+  checkTypeAndDest(ConfigStoredType::Boolean);
+  b = _b;
+  type = ConfigStoredType::Boolean;
+  return true;
+}
+
+bool ConfigValue::set(double _d) {
+  checkTypeAndDest(ConfigStoredType::Double);
+  d = _d;
+  type = ConfigStoredType::Double;
+  return true;
+}
+bool ConfigValue::set(const char* _c) {
+  if (checkTypeAndDest(ConfigStoredType::String)) {
+    s = std::string(_c);
+  } else {
+    new (&s) std::string(_c);
+    type = ConfigStoredType::String;
+  }
+  return true;
+}
+bool ConfigValue::set(const std::string& _s) {
+  if (checkTypeAndDest(ConfigStoredType::String)) {
+    s = _s;
+  } else {
+    new (&s) auto(_s);
+    type = ConfigStoredType::String;
+  }
+  return true;
+}
+
+bool ConfigValue::set(const Magnum::Vector3& _v) {
+  if (checkTypeAndDest(ConfigStoredType::MagnumVec3)) {
+    v = _v;
+  } else {
+    new (&v) auto(_v);
+    type = ConfigStoredType::MagnumVec3;
+  }
+  return true;
+}
+
+bool ConfigValue::set(const Magnum::Quaternion& _q) {
+  if (checkTypeAndDest(ConfigStoredType::MagnumQuat)) {
+    q = _q;
+  } else {
+    new (&q) auto(_q);
+    type = ConfigStoredType::MagnumQuat;
+  }
+  return true;
+}
+
+bool ConfigValue::set(const Magnum::Rad& _r) {
+  if (checkTypeAndDest(ConfigStoredType::MagnumRad)) {
+    r = _r;
+  } else {
+    new (&r) auto(_r);
+    type = ConfigStoredType::MagnumRad;
+  }
+  return true;
+}
+
+bool ConfigValue::checkTypeAndDest(const ConfigStoredType& checkType) {
+  if (type == checkType) {
+    return true;
+  }
+  if (type != ConfigStoredType::Unknown) {
+    // destructor on non-trivial types
+    deleteCurrentValue("checkTypeAndDest function");
+  }
+  return false;
+}
+
+ConfigValue& ConfigValue::operator=(const ConfigValue& otr) {
+  // if current value is string, magnum vector, magnum quat or magnum rad
+  if (type != otr.type) {
+    deleteCurrentValue("assignment operator");
+  }
+  copyValueInto(otr, "assignment operator");
+
+  return *this;
+}
+
+std::string ConfigValue::getAsString() const {
+  switch (type) {
+    case ConfigStoredType::Unknown:
+      return "Undefined value/Unknown type";
+    case ConfigStoredType::Boolean:
+      return (b ? "True" : "False");
+    case ConfigStoredType::Integer:
+      return std::to_string(i);
+    case ConfigStoredType::Double:
+      return std::to_string(d);
+    case ConfigStoredType::String:
+      return s;
+    case ConfigStoredType::MagnumVec3: {
```

Comment:
Doesn't Magnum have some sort of serializer for these types @mosra?

Reply:
It has, one for debug output (where the goal is human readability) and one for (INI-style, space-delimited) configuration values. Neither of those produces what's supposedly needed here, tho (JSON-like formatting).